### PR TITLE
Rework build process for release

### PR DESCRIPTION
**Pull request by mdeltito**
**refactor(build): switch to goreleaser for builds**

This introduces goreleaser to the pipeline in order to more
easily generate multiple signed builds, and distribute assets
to GitHub as a release.

In addition, some tooling changes:

- Reworked the makefile and build process. Everything is now
  run through a docker build container. Local tasks still exist
  for use during development.
- Refactored the test pipeline in Jenkins, removing matrix builds.
  This isn't really necessary since we are distributing pre-built
  binaries.
- Switched from golint (which is deprecated) to golangci-lint. This
  is a more robust code analysis tool.

Ref: LOG-9606

---

**fix(lint): remove dead code and check errors in test**

This fixes two new linting issues:

- Removes unused test functions
- Checks returned errors in httptest server handlers

Ref: LOG-9606

---

**chore(deps): bump terraform-provider-sdk@2.7.0**

This resolves an issue where tests were suddenly failing due
to a broken transient dependency

Ref: LOG-9606

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -24,64 +24,84 @@ pipeline {
...
+      steps {
+        script {
+          withCredentials([
+            string(credentialsId: 'logdna-gpg-key', variable: 'GPG_KEY'),
+            string(credentialsId: 'terraform-provider-servicekey', variable: 'SERVICE_KEY'),
+            string(credentialsId: 'terraform-provider-coveralls', variable: 'COVERALLS_TOKEN')
+          ]) {
+            sh '''
+              set +x
+              echo "$GPG_KEY" > gpgkey.asc
```
Comment (darinspivey) [inline Jenkinsfile:48]: Oh nice

File: Jenkinsfile [hunk 2]
```diff
@@ -24,64 +24,84 @@ pipeline {
...
+      }
 
-        agent {
-          node {
-            label 'ec2-fleet'
-            customWorkspace "${PROJECT_NAME}-${BUILD_NUMBER}-${GO_VERSION}"
+      steps {
+        script {
+          withCredentials([
+            string(credentialsId: 'logdna-gpg-key', variable: 'GPG_KEY'),
```
Comment (darinspivey) [inline Jenkinsfile:42]: Did you have to add this with infra?

File: Makefile [hunk 3]
```diff
@@ -1,44 +1,70 @@
...
+NAME:=logdna
+PROJECT:=terraform-provider-$(NAME)
+VCS_REF:=$(shell git rev-parse --short HEAD)
+BUILD_IMAGE_NAME?=$(PROJECT):$(VCS_REF)
+GOOS:=$(shell go env GOOS)
+GOARCH:=$(shell go env GOARCH)
+
+DOCKER_RUN=docker run --rm -i$(shell [ -t 0 ] && echo t)
+BUILD_ENV=$(DOCKER_RUN) -v $(PWD):/opt/build:Z $(BUILD_FLAGS) $(BUILD_IMAGE_NAME)
+LINT_CMD=$(DOCKER_RUN) -v $(PWD):/app -w /app golangci/golangci-lint:v1.41.1 golangci-lint run -v
```
Comment (darinspivey) [inline Makefile:16]: Looks like this is indeed a better linter!

File: logdna/request_test.go [hunk 4]
```diff
@@ -208,7 +209,8 @@ func TestRequest_MakeRequest(t *testing.T) {
 	t.Run("Handles errors when creating a new HTTP request", func(t *testing.T) {
 		const ERROR = "FAKE ERROR for body reader"
 		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			json.NewEncoder(w).Encode(viewResponse{ViewID: "test123456"})
+			err := json.NewEncoder(w).Encode(viewResponse{ViewID: "test123456"})
+			assert.Nil(err, "No errors")
```
Comment (darinspivey) [inline logdna/request_test.go:213]: Whoops, nice catch.  The new linter?

File: Jenkinsfile (hunk 2 above)
Comment (charlesthomas) [reply Jenkinsfile:42]: if you add `keyFileVariable: 'some_variable_name'` to the `withCredentials` parameters, then Jenkins will put the credentials in a temporary location on the worker, and store the location in the `'some_variable_name'` environment variable. that is probably safer than the echo you're doing a few lines down (although +1 for deleting it in an `always` block).

although if the location is stored in the variable, i'm not sure off the top of my head how you'd inject that in the Dockerfile

Comment (mdeltito) [reply Jenkinsfile:42]: this already existed and matches what the CDN is using

File: logdna/request_test.go (hunk 4 above)
Comment (mdeltito) [reply logdna/request_test.go:213]: yeah just a warning from the new lint suite's static checker

**Comment by mdeltito** [comment]
@charlesthomas @darinspivey I haven't been able to figure out a workaround for running a Jenkins build on a tag, I think we'll need to update the config for this repo (at least ephemerally, at first) to get this going. I can create a REL card for a long-term configuration change for repos that need this.

My plan is to get this merged into the feature branch, update the jenkins project, and then tag a pre-release from there e.g. `v1.1.0-pre` to verify the full process works end-to-end. Note that the current setup in this PR will still _only_ create a draft release, which will require manual publishing to "release". That extra human step is a feature IMO since releases will be automatically pulled by the registry.

**Review by darinspivey — APPROVED** [review]
Looks good, I like the new linter.  As far as the deployment steps, I think you're spot on for us to be ok with a manual release for now.

**Pull request merged**